File: test/benchmark/larceny/gcbench.js
// Larceny Benchmark: gcbench (Node.js)
// GC stress test — binary tree allocation and traversal
'use strict';

function makeTree(depth) {
    if (depth === 0) return { left: null, right: null };
    return { left: makeTree(depth - 1), right: makeTree(depth - 1) };
}

function checkTree(node) {
    if (node.left === null) return 1;
    return 1 + checkTree(node.left) + checkTree(node.right);
}

function main() {
    const minDepth = 4;
    const maxDepth = 14;
    const stretchDepth = maxDepth + 1;

    const stretch = makeTree(stretchDepth);
    process.stdout.write("stretch tree of depth " + stretchDepth + " check: " + checkTree(stretch) + "\n");

    const longLived = makeTree(maxDepth);

    for (let depth = minDepth; depth <= maxDepth; depth += 2) {
        const iterations = 1 << (maxDepth - depth + minDepth);
        let totalCheck = 0;
        for (let i = 0; i < iterations; i++) {
            totalCheck += checkTree(makeTree(depth));
        }
        process.stdout.write(iterations + " trees of depth " + depth + " check: " + totalCheck + "\n");
    }

    process.stdout.write("long lived tree of depth " + maxDepth + " check: " + checkTree(longLived) + "\n");
}

main();

File: test/benchmark/larceny/python/gcbench.py
#!/usr/bin/env python3
# Larceny Benchmark: gcbench (Python)
# GC stress test — binary tree allocation and traversal
import time


def make_tree(depth):
    if depth == 0:
        return (None, None)
    return (make_tree(depth - 1), make_tree(depth - 1))


def check_tree(node):
    if node[0] is None:
        return 1
    return 1 + check_tree(node[0]) + check_tree(node[1])


def main():
    import sys
    sys.setrecursionlimit(100000)

    t0 = time.perf_counter_ns()
    min_depth = 4
    max_depth = 14
    stretch_depth = max_depth + 1

    stretch = make_tree(stretch_depth)
    print(f"stretch tree of depth {stretch_depth} check: {check_tree(stretch)}")

    long_lived = make_tree(max_depth)

    depth = min_depth
    while depth <= max_depth:
        iterations = 1 << (max_depth - depth + min_depth)
        total_check = 0
        for _ in range(iterations):
            total_check += check_tree(make_tree(depth))
        print(f"{iterations} trees of depth {depth} check: {total_check}")
        depth += 2
    t1 = time.perf_counter_ns()

    print(f"long lived tree of depth {max_depth} check: {check_tree(long_lived)}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()
